Firestore/core/src/firebase/firestore/model/transform_operations.cc

```diff
@@ -97,6 +103,11 @@ size_t ArrayTransform::Hash() const {
   return result;
 }
 
+std::string ArrayTransform::ToString() const {
+  const char* name = type_ == Type::ArrayUnion ? "ArrayUnion" : "ArrayRemove";
```

Comment:
Seems reasonable. A few alterations to consider:
1) If you're paranoid, you could: `HARD_ASSERT(type_ == ArrayUnion || type_ == ArrayRemove)` like on line 113.
2) If you're extra paranoid, you could also expand this into a switch, via a TypeToStringIsh() method:
```
const char* TypeToString(type) {
  switch (type) {
    case ArrayUnion: return "ArrayUnion";
    case ArrayRm: return "ArrayRm";
    case Increment: return "Incr";
    case Test: return "Test";
  }
  UNREACHABLE();
}
...
string ArrayTransform::ToSTring() const {
  const char* name = TypeToString(type);
  HARD_ASSERT(type == ArrUn || type == ArrRm, "Unexpected type: %s", name);
  ...
```
This would generate a compile error when we add new types. You could optionally use that new tostring method throughout the other asserts in this file too.
3) If you're super extra paranoid, you could change the hard assert into a switch too:
```
string ArrayTransform::ToSTring() const {
  const char* name = TypeToString(type);
  switch (type) {
    case ArrayUnion: break;
    case ArrayRm: break;
    case Increment:
    case Test:
      HARD_FAIL("Unexpected type %s", name);
  }
  ...
```

All optional.

Reply:
I've added an assertion to the constructor, where the resulting stack will show the code path to create the invalid instance rather than just the stack of an innocent bystander. With that in place I'd prefer not to enforce it here--it's duplicative without adding much value.

Reply:
Adding in the ctor is clearly better. lgtm.

You might consider still adding the TypeToString method to improve the HARD_ASSERT error messages... but that's a separate issue. Still no action required.